x-pack/legacy/plugins/snapshot_restore/public/app/components/policy_delete_provider.tsx

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { Fragment, useRef, useState } from 'react';
+import { EuiConfirmModal, EuiOverlayMask } from '@elastic/eui';
+import { useAppDependencies } from '../index';
+import { deletePolicies } from '../services/http';
+
+interface Props {
+  children: (deletePolicy: DeletePolicy) => React.ReactElement;
+}
+
+export type DeletePolicy = (names: string[], onSuccess?: OnSuccessCallback) => void;
+
+type OnSuccessCallback = (policiesDeleted: string[]) => void;
+
+export const PolicyDeleteProvider: React.FunctionComponent<Props> = ({ children }) => {
+  const {
+    core: {
+      i18n,
+      notification: { toastNotifications },
+    },
+  } = useAppDependencies();
+  const { FormattedMessage } = i18n;
+  const [policyNames, setPolicyNames] = useState<string[]>([]);
+  const [isModalOpen, setIsModalOpen] = useState<boolean>(false);
+  const onSuccessCallback = useRef<OnSuccessCallback | null>(null);
+
+  const deletePolicyPrompt: DeletePolicy = (names, onSuccess = () => undefined) => {
+    if (!names || !names.length) {
+      throw new Error('No policy names specified for deletion');
+    }
+    setIsModalOpen(true);
+    setPolicyNames(names);
+    onSuccessCallback.current = onSuccess;
+  };
+
+  const closeModal = () => {
+    setIsModalOpen(false);
+    setPolicyNames([]);
+  };
+
+  const deletePolicy = () => {
+    const policiesToDelete = [...policyNames];
+    deletePolicies(policiesToDelete).then(({ data, error }) => {
+      const { itemsDeleted, errors } = data || { itemsDeleted: undefined, errors: undefined };
+
+      // Surface success notifications
+      if (itemsDeleted && itemsDeleted.length) {
+        const hasMultipleSuccesses = itemsDeleted.length > 1;
+        const successMessage = hasMultipleSuccesses
+          ? i18n.translate('xpack.snapshotRestore.deletePolicy.successMultipleNotificationTitle', {
+              defaultMessage: 'Deleted {count} policies',
+              values: { count: itemsDeleted.length },
+            })
+          : i18n.translate('xpack.snapshotRestore.deletePolicy.successSingleNotificationTitle', {
+              defaultMessage: "Deleted policy '{name}'",
+              values: { name: itemsDeleted[0] },
+            });
+        toastNotifications.addSuccess(successMessage);
+        if (onSuccessCallback.current) {
+          onSuccessCallback.current([...itemsDeleted]);
+        }
+      }
+
+      // Surface error notifications
+      // `error` is generic server error
+      // `data.errors` are specific errors with removing particular policy(ies)
+      if (error || (errors && errors.length)) {
```

Comment:
I was having trouble following this logic because it contains two parallel logic branches. How about something like this to make the two branches easier to discern and compare?

```js

      function createErrorMessageMultiple(count) {
        return i18n.translate('xpack.snapshotRestore.deletePolicy.errorMultipleNotificationTitle', {
          defaultMessage: 'Error deleting {count} policies',
          values: { count },
        });
      }

      function createErrorMessageSingle(name) {
        return i18n.translate('xpack.snapshotRestore.deletePolicy.errorSingleNotificationTitle', {
          defaultMessage: "Error deleting policy '{name}'",
          values: { name },
        });
      }

      // Surface error notifications
      let errorMessage;

      if (errors.length) {
        // `data.errors` are specific errors with removing particular policy(ies)
        errorMessage = errors.length > 1
          ? createErrorMessageMultiple(errors.length)
          : createErrorMessageSingle(errors[0].name);
      } else if (error) {
        // `error` is generic server error
        errorMessage = policiesToDelete.length > 1
          ? createErrorMessageMultiple(policiesToDelete.length)
          : createErrorMessageSingle(policiesToDelete[0])
      }

      if (errorMessage) {
        toastNotifications.addDanger(errorMessage);
      }
```